### ISSUE TITLE
大佬你好，有几个疑问想问您一下

### ISSUE THREAD

**Issue by shenshaowei** [Chauncey-Jheng/PCRL-MRG#3]
史大哥推荐的论文，拜读了一下，发现是同方向大佬，我有几个疑问可以问一下吗？
1、我下载的https://github.com/tangyuhao2016/CTRG
里的CTRG-Brain-263.rar，发现只有“异常”部分，并且我发现我只有1000多个患者的数据
g的文件下载下来解压完只有500M？是我下载损坏了吗？
2、数据集中提到：“4.1 数据集
我们使用CTRG-Brain（唐等人，2024年）数据集来验证我们模型的性能。该数据集包含6000个样本，总计160,336张CT图像和6000份中文医学报告。按照主流划分方法（史等人，2023年；张等人，2023年），我们将数据集按7:1:2的比例划分为训练集、验证集和测试集。为了保持一致的处理，我们根据专家知识将脑部CT图像样本分为8层，每层包含连续的3张CT图像，为每个样本分配24张CT图像”
能否更详细一些，数据集似乎包括了“正常”和“异常”部分，这部分您是怎么划分的？正常和异常都是7-2-1吗，能否提供一下划分的数据标准（训练集、验证集和测试集具体的患者数）
![Image](https://github.com/user-attachments/assets/47cb4a7c-34fd-4757-8ca9-64a622c775d1)

![Image](https://github.com/user-attachments/assets/d3729207-299d-4e64-b8c0-629588cd79a1)

**Comment by Chauncey-Jheng** [comment]
Thank you for your interest and questions.

1. We encountered no issues during decompression. We recommend using WinRAR for extracting the files to ensure compatibility and stability.


2. The 24 selected images were sampled from the original dataset under the guidance of experienced radiologists. For details on the initial selection strategy, please refer to:
Zhang, Xiaodan, et al. "Weakly guided attention model with hierarchical interaction for brain CT report generation." Computers in Biology and Medicine 167 (2023): 107650.


3. We can provide:
The JSON file listing the 24 selected images.
The JSON files corresponding to the training/validation/test split in a 7:2:1 ratio.
➤ [Download label_jsons.zip](https://github.com/user-attachments/files/20442192/label_jsons.zip)

4. We exclusively used the abnormal subset of the dataset, which contains 6,001 samples in total. For batching purposes, we utilized the first 6,000 samples.

![Image](https://github.com/user-attachments/assets/78280212-59ee-4eab-97d5-f038e1ca3016)

**Comment by shenshaowei** [comment]
> Thank you for your interest and questions.
> 
> 1. We encountered no issues during decompression. We recommend using WinRAR for extracting the files to ensure compatibility and stability.
> 2. The 24 selected images were sampled from the original dataset under the guidance of experienced radiologists. For details on the initial selection strategy, please refer to:
>    Zhang, Xiaodan, et al. "Weakly guided attention model with hierarchical interaction for brain CT report generation." Computers in Biology and Medicine 167 (2023): 107650.
> 3. We can provide:
>    The JSON file listing the 24 selected images.
>    The JSON files corresponding to the training/validation/test split in a 7:2:1 ratio.
>    ➤ [Download label_jsons.zip](https://github.com/user-attachments/files/20442192/label_jsons.zip)
> 4. We exclusively used the abnormal subset of the dataset, which contains 6,001 samples in total. For batching purposes, we utilized the first 6,000 samples.
> 
> ![Image](https://github.com/user-attachments/assets/78280212-59ee-4eab-97d5-f038e1ca3016)

收到，感谢回复，我还有两个疑问
第一，因为每个患者的输入图像都不一定，这部分你们选择24张是怎么选择的，前24张？还是医生挑选的有用的24张。
第二，7-2-1，是按顺序选的吗，6000按顺序分为4200训练集（0-4199）、1200验证集（4200-）和600测试集
按您的意思是总样本数：6001 ，实际使用：前 6000 个样本， 第 6001 个样本被排除
也就是
训练集：4200 条（0 - 4199）
验证集：1200 条（4200 - 5399）
测试集：600 条（5400 - 5999）
对吗？
我将在论文中引用您的论文，感谢

**Comment by Chauncey-Jheng** [comment]
1. Under the guidance of radiologists, we selected 3 images for each of the eight clinical aspects of brain CT, resulting in a total of 24 images.

2. We randomly shuffled the data before splitting it into training, validation, and test sets.
 
3. The json files that we used can be found at: https://github.com/user-attachments/files/20442192/label_jsons.zip.

**Comment by shenshaowei** [comment]
太感谢了

**Comment by shenshaowei** [comment]
hi，我发现，提供的json文件是：
**训练集样本数: 4200
验证集样本数: 601
测试集样本数: 1200**
但是，您说实际使用：前 6000 个样本，但id 6000（即第6001个样本划分在test），似乎对不上，删除了第6001个样本的话，变成：
**训练集样本数: 4200
验证集样本数: 601
测试集样本数: 1199**
请问一下，这对吗

**Comment by shenshaowei** [comment]
另外，论文中的F1，在代码中似乎没有提到如何计算，这部分可以提供一下源码吗？

**Comment by Chauncey-Jheng** [comment]
The F1 score for the Clinical Evaluation (CE) metric is calculated based on the presence of 24 critical pathological terms identified by brain CT radiology experts. [](url)
[ '斑片', '增厚', '出血', '圆形', '钙化', '积血', '脑脊液','颞叶','水肿带','顶叶','受压','右移','左移','筛窦','蝶窦','环形','大脑镰','肿胀','侧脑室后角','小脑半球','桥脑','脑组织','出血灶', '中脑']
| Plaque                              | Thickening            | Bleeding       | Roundness    | Calcification      | Hematuria   |
| Cerebrospinal Fluid                 | Temporal Lobe         | Edema Zone     | Top Lobe     | Compression        | Right Shift |
| Left Shift                          | ethmoid sinus         | sphenoid sinus | ring         | cerebral falx      | swelling    |
| posterior horn of lateral ventricle | cerebellar hemisphere | pontine        | brain tissue | hemorrhagic lesion | midbrain    |
```python
import json
import numpy as np
from sklearn.metrics import precision_score, recall_score, f1_score

key_words = [ '斑片', '增厚', '出血', '圆形', '钙化', '积血', '脑脊液','颞叶','水肿带','顶叶','受压','右移','左移',
        '筛窦','蝶窦','环形','大脑镰','肿胀','侧脑室后角','小脑半球','桥脑','脑组织','出血灶', '中脑']

key_words = list(set(key_words))


def evaluate_from_json(json_file, key_words):
    with open(json_file, encoding='utf-8') as f:
        data = json.load(f)

    total = len(data)
    correct = 0
    p_sum = r_sum = f1_sum = 0

    for sample in data:
        pred_text = sample['prediction']
        gt_text = sample['ground_truth']

        pred_tag = np.array([1 if kw in pred_text else 0 for kw in key_words])
        gt_tag = np.array([1 if kw in gt_text else 0 for kw in key_words])

        p_sum += precision_score(gt_tag, pred_tag, average='binary', zero_division=1)
        r_sum += recall_score(gt_tag, pred_tag, average='binary', zero_division=1)
        f1_sum += f1_score(gt_tag, pred_tag, average='binary', zero_division=1)

        correct += (pred_tag == gt_tag).sum()

    return {
        'acc': correct / (total * len(key_words)),
        'p': p_sum / total,
        'r': r_sum / total,
        'f1': f1_sum / total
    }


if __name__ == '__main__':
    json_path = "your_data.json"

    result = evaluate_from_json(json_path, key_words)

    print(f"  Accuracy: {result['acc']:.4f}")
    print(f"  Precision: {result['p']:.4f}")
    print(f"  Recall: {result['r']:.4f}")
    print(f"  F1-Score: {result['f1']:.4f}")

```

> 另外，论文中的F1，在代码中似乎没有提到如何计算，这部分可以提供一下源码吗？

**Comment by Chauncey-Jheng** [comment]
We used drop_last when constructing the dataloader, so the last incomplete batch will be automatically skipped when building the eval_dataloader.

> hi，我发现，提供的json文件是： **训练集样本数: 4200 验证集样本数: 601 测试集样本数: 1200** 但是，您说实际使用：前 6000 个样本，但id 6000（即第6001个样本划分在test），似乎对不上，删除了第6001个样本的话，变成： **训练集样本数: 4200 验证集样本数: 601 测试集样本数: 1199** 请问一下，这对吗
